### ISSUE TITLE
Replace TimeoutCancellationException with TimeoutException

### ISSUE THREAD

**Issue by qwwdfsad** [Kotlin/kotlinx.coroutines#1374]
In the current version of the library, `withTimeout` throws `TimeoutCancellationException` when a timeout was exceeded.

It can lead to very subtle errors, for example:
```
launch {
    val result = withTimeout(...) {
        // ... some computation ...
    }
    // process result
}
```

`TimeoutCancellationException` is `CancellationException`, thus is never reported.
But in the snippet above, it's likely to be a programmatic error. If it is expected to miss the deadline, then `withTimeoutOrNull` should be used **explicitly**.

**Comment by artbez** [comment]
Hi, @qwwdfsad !
My team and I also faced this problem. Do you have any expectations about the time this task will be solved?

**Comment by qwwdfsad** [comment]
Could you please elaborate on the problem you've faced?

**Comment by artbez** [comment]
Sure, we use `withTimeout` for restricting our inner functions with deadlines. So, any exceptions except CancellationException will be reported from our business logic code, but TimeoutCancellationException won’t. We want to parse such exceptions as others. And, as you said before, it would be possible if withTimeout threw TimeoutException instead of TimeoutCancellationException. Another case is when we use launch within coroutineScope. Exceded deadline in one job doesn't cancel other jobs. Of course, we can use our own custom launch which catches TimeoutCancellationException inside and throws TimeoutException or we can use withTimeoutOrNull. But, we'd like to know if you plan to change the behavior of withTimeout function. 

Thank you!

**Comment by qwwdfsad** [comment]
The preliminary design decision is to deprecate `withTimeout` function with `@LowPriorityInOverloadResolution` and re-introduce it in the brand new package `kotlinx.coroutines.time.*` with the same name.

The change is now blocked by the IDE bug that doesn't replace a signature to the same name in a different package

**Comment by dkhalanskyjb** [comment]
This could be a good time to also avoid this behavior https://github.com/Kotlin/kotlinx.coroutines/issues/1914 by checking at the end of the block if the coroutine was canceled and throwing in that case.

**Comment by qwwdfsad** [comment]
The same applies to the cancellation before `withTimeout` even start (that is also observed in #1914).

Historically, it was on par with `withContext` behaviour. Though the latter actively evolved (#962, #785, then #791), while `withTimeout` has been left as is. Its behaviour is clearly lagging behind what is expected from it by default and is worth changing.

It seems like we already have three potential changes in the function semantics:

* Exception type (more important, it's CE -> non-CE)
* Non-atomic cancellation (it also will help to get rid of all the hacks in our `undispatchedResult` and `startUndispatchedOrReturnIgnoreTimeout`) on completion 
* Avoid entrance into the body of `withTimeout` if timeout already kicked in before coroutine start

**Comment by qwwdfsad** [comment]
It can be the case that the behaviour is too breaking and thus we should come up with a better name instead of silently shadowing the original signature. 
What concerns me is the fact that timeouts are non-deterministic and are likely to happen under the load or if things go south, meaning that users are likely to learn about this change from production crashes rather than regular testing

**Comment by jQrgen** [comment]
Relevant: https://betterprogramming.pub/the-silent-killer-thats-crashing-your-coroutines-9171d1e8f79b

**Comment by wkornewald** [comment]
> Relevant: https://betterprogramming.pub/the-silent-killer-thats-crashing-your-coroutines-9171d1e8f79b

The article correctly describes some root causes of the problem, i.e. CancellationException being used in places where an error-indicating exception would be needed instead.

However the article's proposed solution is to treat all abortable cancellations as errors (i.e. where the job wasn't canceled). This can't be correct. Just because >90% of CancellationExceptions use Job cancellation doesn't mean you can turn all other cancellations into errors. Ignoring those abortable cases with "but you shouldn't be catching errors there anyway" or "So far, I haven’t found any major issues" is a shaky argument (and in our codebase it would quickly hit issues). The article's solution is a workaround that breaks fundamental "safe bubbling" behavior of cancellations where you don't want to trigger error handling (like AbortFlowException).

The real solution is to fix the incorrect usages of CancellationException or at least provide alternative APIs where breaking changes wouldn't be acceptable. The bubbling behavior should not be broken.

**Comment by jQrgen** [comment]
@wkornewald, thanks for the clarification.

**Comment by joffrey-bion** [comment]
Is there a similar plan to "fix" `Deferred.await()` / `SendChannel.send()` / `ReceiveChannel.receive()` in the same way? i.e. only throw `CancellationException` when the calling coroutine is cancelled, but NOT in situations where the producer/consumer on the other end is cancelled (in which case some other exception that's not a subtype of `CancellationException` would be more appropriate, in order to avoid cancelling the calling coroutine silently)

**Comment by dkhalanskyjb** [comment]
We are yet to decide what to do about this. See https://github.com/Kotlin/kotlinx.coroutines/issues/3658

**Comment by Legion2** [comment]
Also facing this inconsistency, as a workaround I now use `withTimeoutOrNull` and throw an exception on null value. Took my a long time to figure out why my code got silently canceled instead of getting a real exception which explains the timeout.

**Comment by masc3d** [comment]
this will also happen when combining flow operators `timeout` and `merge`. very confusing and difficult to track down in  complex flows.

**Comment by scherrsasrf** [comment]
Just wasted 4 hours tracking down jobs that were cancelled on timeout but didn't throw an exception that could be logged. We are now using something like this: 
```kotlin
class TimeoutException(cause: Exception) : Exception(cause)

suspend fun <T> withDeadline(timeout: Duration, block: suspend () -> T): T = try {
    withTimeout(timeout) { block() }
} catch (e: TimeoutCancellationException) {
    throw TimeoutException(e)
}
```

**Comment by eygraber** [comment]
I knew about this issue and still got bit by it.

**Comment by qwwdfsad** [comment]
Here goes the discussion for the potential solution: https://github.com/Kotlin/kotlinx.coroutines/pull/4356
